File: internal/content/testdata/bench/java/generate-balanced-parentheses.java
import java.util.*;

class Solution {
    public String[] generate_parentheses(long n) {
        List<String> results = new ArrayList<>();
        backtrack(results, "", 0, 0, (int) n);
        return results.toArray(new String[0]);
    }

    private void backtrack(List<String> results, String s, int openCount, int closeCount, int n) {
        if (s.length() == 2 * n) {
            results.add(s);
            return;
        }
        if (openCount < n) backtrack(results, s + "(", openCount + 1, closeCount, n);
        if (closeCount < openCount) backtrack(results, s + ")", openCount, closeCount + 1, n);
    }
}

File: internal/content/testdata/bench/rust/generate-balanced-parentheses.rs
fn generate_parentheses(n: i64) -> Vec<String> {
    fn backtrack(s: String, open_count: i64, close_count: i64, n: i64, results: &mut Vec<String>) {
        if s.len() as i64 == 2 * n {
            results.push(s);
            return;
        }
        if open_count < n {
            let mut t = s.clone();
            t.push('(');
            backtrack(t, open_count + 1, close_count, n, results);
        }
        if close_count < open_count {
            let mut t = s.clone();
            t.push(')');
            backtrack(t, open_count, close_count + 1, n, results);
        }
    }
    let mut results: Vec<String> = Vec::new();
    backtrack(String::new(), 0, 0, n, &mut results);
    results
}
